compiler/rustc_typeck/src/check/pat.rs

```diff
@@ -1040,6 +1042,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 vec![(left, "(".to_string()), (right.shrink_to_hi(), ")".to_string())],
                 Applicability::MachineApplicable,
             );
+        } else if fields.len() > subpats.len() {
+            let after_fields_span = if pat_span == DUMMY_SP {
+                pat_span
+            } else {
+                pat_span.with_hi(pat_span.hi() - BytePos(1)).shrink_to_hi()
+            };
+
+            let mut wildcard_sugg =
+                iter::repeat("_").take(fields.len() - subpats.len()).collect::<Vec<_>>().join(", ");
+            if !subpats.is_empty() {
+                wildcard_sugg = String::from(", ") + &wildcard_sugg;
+            }
+
+            let rest_sugg = if subpats.is_empty() { "..".to_owned() } else { ", ..".to_owned() };
+
+            err.span_suggestion(
+                after_fields_span,
+                "use `_` to explicitly ignore each field",
+                wildcard_sugg,
+                Applicability::MaybeIncorrect,
+            );
+            err.span_suggestion(
+                after_fields_span,
+                "use `..` to ignore all unmentioned fields",
+                rest_sugg,
+                Applicability::MaybeIncorrect,
+            );
```

Comment:
I believe you might want to use `span_suggestions` to suggest multiple items, rustfix checks this but maybe not in our case. https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic/struct.Diagnostic.html#method.span_suggestions https://github.com/rust-lang/rustfix/issues/78

```suggestion
            err.span_suggestions(
                after_fields_span,
                "use `_` to explicitly ignore each field or `..` to ignore all unmentioned fields",
                vec![wildcard_sugg, rest_sugg].into_iter(),
                Applicability::MaybeIncorrect,
            );
```
I wonder if `vec![wildcard_sugg, rest_sugg].into_iter(),` could be improved?

But we can even leave a `FIXME` mentioning where it could improve the applicability to a more correct one with some type matching to know which position to fill with `_`.

Reply:
I like this idea, but unfortunately I don't think the loss in readability is worth it. Currently, the output looks like this:

```
help: use `_` to explicitly ignore each field
   |
LL |         E::S(_, _) => {}
   |              ^^^^
help: use `..` to ignore all unmentioned fields
   |
LL |         E::S(..) => {}
   |              ^^
```

But using your suggestion makes the output look like this:

```
help: use `_` to explicitly ignore each field or `..` to ignore all unmentioned fields
   |
39 |         E::S(_, _) => {}
   |              ^^^^
39 |         E::S(..) => {}
   |              ^^
```

Which is a bit overwhelming and not as clear for the user. Perhaps we can make changes to the compiler's error system to make multiple suggestions easier to read for the user.

There is precedent for using multiple calls to `span_suggestion` to make the output easier to read:

```
error: unexpected `,` in pattern
 --> src/main.rs:2:10
  |
2 |     let x, y = 3;
  |          ^
  |
help: try adding parentheses to match on a tuple...
  |
2 |     let (x, y) = 3;
  |         ^^^^^^
help: ...or a vertical bar to match on multiple alternatives
  |
2 |     let x | y = 3;
  |         ^^^^^
```

Reply:
> Which is a bit overwhelming and not as clear for the user. Perhaps we can make changes to the compiler's error system to make multiple suggestions easier to read for the user.

Yes, this is a better option but we should make rustfix aware of this, if you add it please add it to rustfix too.